### PR TITLE
feat: Leverage Operations Test

### PR DESCRIPTION
**Pull request by RafilxTenfen**
## Description

- Implemented operations test functions of `x/leverage/simulation` module

ref: #354  

----

### Author Checklist

*All items are required. Please add a note to the item if the item is not applicable and
please add links to any relevant follow up issues.*

I have...

- [x] included the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [x] added appropriate labels to the PR
- [ ] added `!` to the type prefix if API or client breaking change
- [x] targeted the correct branch (see [PR Targeting](https://github.com/umee-network/umee/blob/main/CONTRIBUTING.md#pr-targeting))
- [x] provided a link to the relevant issue or specification
- [ ] added a changelog entry to `CHANGELOG.md`
- [x] included comments for [documenting Go code](https://blog.golang.org/godoc)
- [ ] updated the relevant documentation or specification
- [ ] reviewed "Files changed" and left comments if necessary
- [ ] confirmed all CI checks have passed

### Reviewers Checklist

*All items are required. Please add a note if the item is not applicable and please add
your handle next to the items reviewed if you only reviewed selected items.*

I have...

- [ ] confirmed the correct [type prefix](https://github.com/commitizen/conventional-commit-types/blob/v3.0.0/index.json) in the PR title
- [ ] confirmed `!` in the type prefix if API or client breaking change
- [ ] confirmed all author checklist items have been addressed
- [ ] reviewed state machine logic
- [ ] reviewed API design and naming
- [ ] reviewed documentation is accurate
- [ ] reviewed tests and test coverage
- [ ] manually tested (if applicable)

### CODE REVIEW

File: x/leverage/types/tx.go
Comment (alexanderbez) [inline x/leverage/types/tx.go:87]: Why did we have to implement `Route` and `Type` for these message types? IIRC, these are for the old deprecated means of handling messages (pre gRPC routing).

File: x/leverage/simulation/operations.go
Comment (alexanderbez) [inline x/leverage/simulation/operations.go:178]: I think the convention we use is `uToken`, right @toteki?

```suggestion
		from, uToken, skip := randomCollateralFields(r, ctx, accs, lk)
```

**Review by alexanderbez — COMMENTED** [review]
Excellent work @RafilxTenfen

File: x/leverage/simulation/operations.go
Comment (toteki) [reply x/leverage/simulation/operations.go:178]: yeah, usually token and uToken for generic `sdk.Coin` values

File: x/leverage/types/tx.go
Comment (RafilxTenfen) [reply x/leverage/types/tx.go:87]: If we don't implement that it fails to Marshal a NewOperationMsg at `cosmos-sdk/types/simulation/types.go:84`

File: x/leverage/simulation/operations.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations.go]: Update coin to token at https://github.com/umee-network/umee/pull/433/commits/c210c5605b3fe23e67fec59d8b991efc35a91361

File: x/leverage/types/tx.go
Comment (alexanderbez) [reply x/leverage/types/tx.go:87]: I see. That's unfortunate :(

**Review by alexanderbez — APPROVED** [review]

**Comment by alexanderbez** [comment]
utACK. @toteki can you also give this a review 🙏

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:369]: Since these are collateral `uTokens` and not borrowed `tokens`, consider calling them `rewardTokens` instead of `repaymentTokens`. Liquidate repays borrowed `tokens` and receives collateral `uTokens` as a reward.

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:358]:
```suggestion
// randomLiquidateFields returns two random accounts to be used as a liquidator
// and a borrower in a MsgLiquidate transaction, as well as a random sdk.Coin from
// the borrower's collateral. It returns skip=true if no collateral is found.
```

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:342]: // randomCollateralFields returns a random account and an sdk.Coin from its collateral.
// It returns skip=true if no collateral is found.

Comment (toteki) [inline x/leverage/simulation/operations.go:342]:
```suggestion
// randomCollateralFields returns a random account and an sdk.Coin from its collateral.
// It returns skip=true if no collateral was found.
```

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:326]:
```suggestion
// randomSpendableFields returns a random account and an sdk.Coin from its spendable
// coins. It returns skip=true if the account has zero spendable coins.
```

File: x/leverage/simulation/operations.go
Comment (toteki) [reply x/leverage/simulation/operations.go:369]: Additional note, after seeing how it's used:

it looks like this function needs one more return - specifically, liquidation requires a `repaymentTokens` from `GetBorrowerBorrows(borrower.Address)` and a `rewardDenom` from `GetBorrowerCollateral(borrower.Address)`.

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:252]: When simulating `MsgRepay`, the coin to repay should be a subset of `GetBorrowerBorrows(borrower.Address)`

I recommend creating a function `randomBorrowedFields` or something similar

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:146]: There are two places to find coins eligible for withdrawal
1) A user's collateral-enabled uTokens, which are searched here using `GetBorrowerCollateral`
2) A user's collateral-_disabled_ uTokens, which are part of their `SpendableCoins`. In order to find only uTokens in `SpendableCoins` and ignore tokens, you must look for denoms that start with `"u/"`

In order to simulate withdrawals of both types, a function `randomWithdrawFields` might be created which collects both types of uTokens eligible for withdrawal into a single `sdk.Coins`, and returns `simtypes.RandSubsetCoins` from that.

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:178]: When borrowing assets, a user can borrow any accepted denom, not just those whose uTokens they have in their collateral.

This function should therefore select a random `Token.BaseDenom` from `lk.GetAllRegisteredTokens` and then generate a random amount.

File: x/leverage/simulation/operations.go
Comment (toteki) [inline x/leverage/simulation/operations.go:218]: We can generate the denom here using `lk.FromTokenToUTokenDenom` on a random `Token.BaseDenom` from `lk.GetAllTegisteredTokens`.

In that way, we are selecting from all possible uToken denominations.

File: x/leverage/simulation/operations_test.go
Comment (toteki) [inline x/leverage/simulation/operations_test.go:48]:
```suggestion
```
This would be a `token:uToken` exchange rate, rather than an oracle price, so it should not need to be set. If it must be set though, make sure it is greater than 1.0

File: x/leverage/simulation/operations_test.go
Comment (toteki) [inline x/leverage/simulation/operations_test.go:36]: Let's create three different accepted assets and give them different parameters and oracle prices, for maximum chaos.

(Notably, so the simulation can test liquidations where the repay and reward denom are different, and other such scenarios)

File: x/leverage/simulation/operations_test.go
Comment (toteki) [inline x/leverage/simulation/operations_test.go:59]: After adding multiple accepted assets, also be sure to mint some of every type to the accounts.

**Review by toteki — CHANGES_REQUESTED** [review]
We should add multiple accepted asset types, and ensure that we are selecting message inputs from the right places (e.g. all asset types, borrowed only, collateral only) before merging this

File: x/leverage/simulation/operations.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations.go]: done at https://github.com/umee-network/umee/pull/433/commits/e0d03fa4126c475652d58d36906eeffe8ee33993

File: x/leverage/simulation/operations.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations.go]: done at https://github.com/umee-network/umee/pull/433/commits/2a4fdb2d124bb3c326aa5e5f8593b7967b8fee2b

File: x/leverage/simulation/operations.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations.go]: done at https://github.com/umee-network/umee/pull/433/commits/6f97a9268077f31aed9f5185b9287a178c935ffd

File: x/leverage/simulation/operations_test.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations_test.go]: done at https://github.com/umee-network/umee/pull/433/commits/9da105c9fdc1483edca89ca2b632fb333c21b56f

File: x/leverage/simulation/operations_test.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations_test.go:36]: done at https://github.com/umee-network/umee/pull/433/commits/fe4fa741d42ca44780be0911235763b174e3bf4f

File: x/leverage/simulation/operations_test.go
Comment (RafilxTenfen) [reply x/leverage/simulation/operations_test.go:84]: done at https://github.com/umee-network/umee/pull/433/commits/fe4fa741d42ca44780be0911235763b174e3bf4f

**Review by toteki — APPROVED** [review]
Added a few more commits, approving now

**Pull request merged**
